AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMMs/DefaultCMM.dfy

```diff
@@ -478,25 +480,36 @@ module DefaultCMM {
       var requiredEncryptionContextKeys := [];
       if input.reproducedEncryptionContext.Some? {
         var keysSet := input.reproducedEncryptionContext.value.Keys;
-        while keysSet != {}
+        ghost var keysSet' := keysSet;
+        var keysSeq := SortedSets.ComputeSetToSequence(keysSet);
+        var i := 0;
+        while i < |keysSeq|
+          invariant Seq.HasNoDuplicates(keysSeq)
+          invariant forall j | 0 <= j < i && i < |keysSeq| :: keysSeq[j] !in keysSet'
+          invariant forall j | i <= j < |keysSeq| :: keysSeq[j] in keysSet'
+          invariant |keysSet'| == |keysSeq| - i
           invariant forall key
                       |
-                      && key in input.reproducedEncryptionContext.value
+                      && key in input.reproducedEncryptionContext.value.Keys
                       && key in input.encryptionContext
-                      && key !in keysSet
-                      :: input.reproducedEncryptionContext.value[key] == input.encryptionContext[key]
+                      && key !in keysSet'
+                      :: input.encryptionContext[key] == input.reproducedEncryptionContext.value[key]
```

Comment:
Slight nit, adjusting the verification makes it a little hard to verify the lack of change :)
This way would have made it clear that the only actual change was to change the invariant to use the ghost value.
and that everything else was additions.

 ```dafny
                      && key in input.reproducedEncryptionContext.value
                      && key in input.encryptionContext
                      && key !in keysSet'
                      :: input.reproducedEncryptionContext.value[key] == input.encryptionContext[key]
```

Reply:
Good point. Those changes are an artifact of attempting to reduce brittleness, before I realized that it was actually missing information leading to failure here. They're not necessary, and I'd be happy to undo them.